File: add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        
        ListNode dummy = new ListNode(-1);
        ListNode curr = dummy;
        
        ListNode t1 = l1;
        ListNode t2 = l2;

        int carry = 0;
        // int sum = 0;

        while (t1 != null || t2 != null) {
            int sum = carry;
            if (t1 != null) {
                sum += t1.val;
                t1 = t1.next;
            }

            if (t2 != null) {
                sum += t2.val;
                t2 = t2.next;
            }

            // need to undertand this logic better
            ListNode newNode = new ListNode(sum % 10); 
            carry = sum / 10; 

            curr.next = newNode;
            curr = curr.next;

            // if (t1 != null) {  }

            // if (t2 != null){  }

        }

        if (carry != 0) {
            ListNode newNode = new ListNode(carry);
            curr.next = newNode;
        }

        return dummy.next;
    }
}

File: 2-add-two-numbers/add-two-numbers.swift
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init() { self.val = 0; self.next = nil; }
 *     public init(_ val: Int) { self.val = val; self.next = nil; }
 *     public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
 * }
 */
class Solution {
    func addTwoNumbers(_ l1: ListNode?, _ l2: ListNode?) -> ListNode? {
        
        var carry = 0
        let dummy = ListNode(-1)
        var curr = dummy

        var l1 = l1, l2 = l2 // Mutable copies to traverse the lists

        while l1 != nil || l2 != nil || carry != 0 {
            let x = l1?.val ?? 0
            let y = l2?.val ?? 0

            var sum = x + y + carry
            carry = sum / 10
            curr.next = ListNode(sum % 10)
            curr = curr.next!

            if l1 != nil { l1 = l1?.next }
            if l2 != nil { l2 = l2?.next }
        }

        return dummy.next
    }
}